<a href="https://colab.research.google.com/github/josedanielisidororeyes/Advanced_Data_Engineering/blob/main/Ejercicio_Representaci%C3%B3n_Texto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Nombre del Alumno: José Daniel Isidoro Reyes
#Matrícula: 261552
#Nombre de la Tarea: Ejercicios de representación de texto
#Fecha: 06/03/2026


In [43]:
# Carga de librerías

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import nltk
from nltk.corpus import stopwords
import pandas  as pd

# Ejercicio 1

## Creación de CountVectirizer y TdifVectorizer

In [12]:
corpus_es = [
    "Me encantó la película, los actores fueron increíbles",
    "No me gustó la película, el guion fue muy malo",
    "La actuación fue excelente, pero la historia era predecible",
    "Película aburrida, me dormí a la mitad",
    "¡Maravillosa! Recomiendo esta película a todos mis amigos",
    "El guion estaba mal escrito, pero la actuación salvó la película",
    "Demasiado larga y lenta, no la volvería a ver",
    "Me gustó mucho, los efectos especiales fueron impresionantes",
    "No es mala, pero esperaba algo más emocionante",
    "Una obra maestra del cine español, increíble dirección",
    "El final fue confuso, no entendí nada",
    "Excelente cinematografía, pero el guion flojo",
    "Demasiado predecible, ya sabía lo que iba a pasar",
    "Me reí mucho, muy divertida y entretenida",
    "La música fue espectacular, pero los actores no convencieron",
    "Película mediocre, no aporta nada nuevo",
    "Excelente historia, emocionante hasta el final",
    "No me gustó la ambientación, parecía de bajo presupuesto",
    "Muy buena dirección y fotografía, la recomiendo",
    "El guion tenía agujeros, pero los efectos visuales impresionan",
    "Aburrida, diálogo poco natural",
    "Gran actuación de los protagonistas, realmente me impactó",
    "El ritmo es lento, pero la historia es interesante",
    "No la recomiendo, desperdicié mi tiempo",
    "Un clásico moderno, me encantó cada escena",
    "Las escenas de acción fueron espectaculares",
    "El humor es pobre y los personajes poco creíbles",
    "Me emocioné, la historia me llegó al corazón",
    "La trama es confusa y difícil de seguir"
]

In [7]:
# Creación de CountVectorizer sin Procesamiento
vectorizer_count  =  CountVectorizer()
X_count  =  vectorizer_count.fit_transform(corpus_es)

In [11]:
# Creación de TfidfVectorizer sin Procesamiento
vectorizer_tfidf =  TfidfVectorizer()
X_xfidf =  vectorizer_tfidf.fit_transform(corpus_es)

In [14]:
# Impresión de resultados
print("---Ejercicio 1 ---")
print(f"Dimensiones de CountVectorizer: {X_count.shape}")
print(f"Dimensiones de TfidfVectorizer: {X_xfidf.shape}")

---Ejercicio 1 ---
Dimensiones de CountVectorizer: (29, 124)
Dimensiones de TfidfVectorizer: (29, 124)


In [17]:
# Impresión de los primeros 10 tokens con sus respectivos índices
tokens_count  =  list(vectorizer_count.vocabulary_.items())[:10]
print(f"\nPrimeros 10 tokens (Count): {tokens_count}")


Primeros 10 tokens (Count): [('me', 81), ('encantó', 34), ('la', 68), ('película', 98), ('los', 75), ('actores', 2), ('fueron', 54), ('increíbles', 66), ('no', 93), ('gustó', 57)]


In [20]:
# Diferencia entre ambas tablas

In [22]:
# Obtención dle nombre de las palabras
tokens  =  vectorizer_count.get_feature_names_out()

df_count  =  pd.DataFrame(X_count.toarray(), columns  =  tokens)
df_tfidf =  pd.DataFrame(X_xfidf.toarray(), columns  =  tokens)

# Impresión de primeras filas
print("--- CountVectorizer ---")
print(df_count.head())

print("\n--- TFIDF ---")
print(df_tfidf.head())

--- CountVectorizer ---
   aburrida  acción  actores  actuación  ...  ver  visuales  volvería  ya
0         0       0        1          0  ...    0         0         0   0
1         0       0        0          0  ...    0         0         0   0
2         0       0        0          1  ...    0         0         0   0
3         1       0        0          0  ...    0         0         0   0
4         0       0        0          0  ...    0         0         0   0

[5 rows x 124 columns]

--- TFIDF ---
   aburrida  acción   actores  actuación  ...  ver  visuales  volvería   ya
0  0.000000     0.0  0.412117     0.0000  ...  0.0       0.0       0.0  0.0
1  0.000000     0.0  0.000000     0.0000  ...  0.0       0.0       0.0  0.0
2  0.000000     0.0  0.000000     0.3453  ...  0.0       0.0       0.0  0.0
3  0.458233     0.0  0.000000     0.0000  ...  0.0       0.0       0.0  0.0
4  0.000000     0.0  0.000000     0.0000  ...  0.0       0.0       0.0  0.0

[5 rows x 124 columns]


<p align  =  "Justify "

Como se puede observar CountVectorizer es una matriz que muestra el número de veces que una palabra aparece en el documento. En cambio, TF-IDF muestra el peso (importancia relativa) de cada palabra en el documento.

# Ejercicio 2

In [35]:
nltk.download('stopwords')
stopwords_español  =  stopwords.words('spanish')

# CountVectorizer con Procesamiento de Datos
vectorizer_count  =  CountVectorizer(
    lowercase =  True,
    strip_accents =  'unicode',
    stop_words =  stopwords_español
)
X_count  =  vectorizer_count.fit_transform(corpus_es)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
/usr/local/lib/python3.12/dist-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['eramos', 'estabamos', 'estais', 'estan', 'estara', 'estaran', 'estaras', 'estare', 'estareis', 'estaria', 'estariais', 'estariamos', 'estarian', 'estarias', 'esteis', 'esten', 'estes', 'estuvieramos', 'estuviesemos', 'fueramos', 'fuesemos', 'habeis', 'habia', 'habiais', 'habiamos', 'habian', 'habias', 'habra', 'habran', 'habras', 'habre', 'habreis', 'habria', 'habriais', 'habriamos', 'habrian', 'habrias', 'hayais', 'hubieramos', 'hubiesemos', 'mas', 'mia', 'mias', 'mio', 'mios', 'seais', 'sera', 'seran', 'seras', 'sere', 'sereis', 'seria', 'seriais', 'seriamos', 'serian', 'serias', 'si', 'tambien', 'tendra', 'tendran', 'tendras', 'tendre', 'tendreis', 'tendria', 'tendriais'

In [38]:
# TF-DF Vectorizer con Procesamiento
vectorizer_tfidf = TfidfVectorizer(
    lowercase =  True,
    strip_accents = 'unicode',
    stop_words =  stopwords_español
)
Xtfidf =  vectorizer_tfidf.fit_transform(corpus_es)

/usr/local/lib/python3.12/dist-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['eramos', 'estabamos', 'estais', 'estan', 'estara', 'estaran', 'estaras', 'estare', 'estareis', 'estaria', 'estariais', 'estariamos', 'estarian', 'estarias', 'esteis', 'esten', 'estes', 'estuvieramos', 'estuviesemos', 'fueramos', 'fuesemos', 'habeis', 'habia', 'habiais', 'habiamos', 'habian', 'habias', 'habra', 'habran', 'habras', 'habre', 'habreis', 'habria', 'habriais', 'habriamos', 'habrian', 'habrias', 'hayais', 'hubieramos', 'hubiesemos', 'mas', 'mia', 'mias', 'mio', 'mios', 'seais', 'sera', 'seran', 'seras', 'sere', 'sereis', 'seria', 'seriais', 'seriamos', 'serian', 'serias', 'si', 'tambien', 'tendra', 'tendran', 'tendras', 'tendre', 'tendreis', 'tendria', 'tendriais', 'tendriamos', 'tendrian', 'tendrias', 'teneis', 'tengais', 'tenia', 'teniais', 'teniamos', 'tenian', 'tenias', 'tuvie

In [40]:
# Impresión de resultados
print("---Ejercicio 1 ---")
print(f"Dimensiones de CountVectorizer: {X_count.shape}")
print(f"Dimensiones de TfidfVectorizer: {Xtfidf.shape}")

# Impresión  de primeros 10 tokens
vocabulario  =  vectorizer_count.get_feature_names_out()
print(f"\nPrimeros 10 tokens: {vocabulario[:10]}")

---Ejercicio 1 ---
Dimensiones de CountVectorizer: (29, 94)
Dimensiones de TfidfVectorizer: (29, 94)

Primeros 10 tokens: ['aburrida' 'accion' 'actores' 'actuacion' 'agujeros' 'ambientacion'
 'amigos' 'aporta' 'bajo' 'buena']


In [42]:

# Comparación de resultados
df_comparativo =  pd.DataFrame({
    'Token': vocabulario,
    'Count': X_count.toarray()[0],
    'TF-IDF': Xtfidf.toarray()[0]
})

print(df_comparativo[df_comparativo['Count'] > 0])

         Token  Count    TF-IDF
2      actores      1  0.512092
29     encanto      1  0.512092
54  increibles      1  0.574963
75    pelicula      1  0.380712


La principal diferencia es la reduccción del numero de dimensiones en el documento (124 vs 94). Adicionalmente, al eliminar stopwords se observa que se conservan tokens con mayor caracter semantica. Finalmente, se observa que la frencuencia realtiva de cada palabra se vuelve más precisa al eliminar stopwords.